tools/src/main/java/org/wildfly/swarm/tools/DeclaredDependencies.java

```diff
@@ -37,62 +40,171 @@
+        List<DependencyTree<ArtifactSpec>> sources = new ArrayList<>();
+        if (includeUnsolved) {
+            sources.add(this);
+        }
+        if (includePresolved) {
+            sources.add(presolvedDependencies);
+        }
+        sources.forEach(s -> s.getDirectDeps().stream()
+                .filter(d -> !isThorntailRunner(d))
+                .forEach(d -> deps.addAll(s.getTransientDeps(d))));
+        return deps;
+    }
+
+    /**
+     * Get the collection of all dependencies associated with this instance (direct & transient) excluding those marked as test.
+     *
+     * @return the collection of all dependencies associated with this instance excluding those marked as test.
+     */
     public Collection<ArtifactSpec> getRuntimeExplicitAndTransientDependencies() {
         Collection<ArtifactSpec> allDeps = new HashSet<>();
-
-        getDirectDeps()
-                .stream()
-                .filter(spec -> !spec.scope.equals("test"))
-                .forEach(spec -> {
-                    allDeps.add(spec);
-                    allDeps.addAll(getTransientDependencies(spec));
-                });
+        Arrays.asList(this, presolvedDependencies).forEach(source -> {
+            source.getDirectDeps()
+                    .stream()
+                    .filter(spec -> !spec.scope.equals("test"))
+                    .forEach(spec -> {
+                        allDeps.add(spec);
+                        allDeps.addAll(source.getTransientDeps(spec));
+                    });
+        });
 
         return allDeps;
     }
 
+    /**
+     * Get the transient dependencies defined for the given artifact specification.
+     *
+     * @param artifact the artifact specification.
+     * @return the transient dependencies defined for the given artifact specification.
+     */
     public Collection<ArtifactSpec> getTransientDependencies(ArtifactSpec artifact) {
-        return getTransientDeps(artifact);
+        Set<ArtifactSpec> deps = new HashSet<>();
+        if (this.isDirectDep(artifact)) {
+            deps.addAll(getTransientDeps(artifact));
+        }
+        if (presolvedDependencies.isDirectDep(artifact)) {
+            deps.addAll(presolvedDependencies.getTransientDeps(artifact));
+        }
+        return deps;
     }
 
     /**
      * 'Presolved' means a build component (i.e. mojo) pre-computed the transient dependencies
      * and thus we can assume this set is fully and correctly resolved
      *
-     * @return
+     * @return true if all the dependencies in this instance are pre-solved, false otherwise.
      */
     public boolean isPresolved() {
-        return getTransientDependencies().size() > 0;
+        // Check if the tool has added unsolved dependencies to this instance that need further resolution.
+        boolean unsolvedDependenciesExist = getDirectDeps().stream().anyMatch(d -> !getTransientDeps(d).isEmpty());
+        return !unsolvedDependenciesExist;
```

Comment:
The original (`getTransientDependencies().size() > 0`) is semantically equivalent, right?

Reply:
Correct. It should return `true` if any of the direct-deps has transient deps.

Reply:
I was just asking myself if this change has some deeper reason :-)